File: 2019/CSharp/Day08.cs
using System.Collections.Generic;
using System.Linq;
using System.Text;
using AdventOfCode.Common.Extensions;
using Xunit;
using Xunit.Abstractions;

namespace AdventOfCode.CSharp.Year2019
{
    /// <summary>
    /// https://adventofcode.com/2019/day/8
    /// </summary>
    public class Day08 : TestBase
    {
        public Day08(ITestOutputHelper output) : base(output, 8) { }

        private List<int> GetInput(string name)
        {
            var input = Input.ReadLines(DAY, name)
                .First()
                .Select(c => (int)char.GetNumericValue(c))
                .ToList();
            return input;
        }

        private int? GetPart1Expected(string name)
        {
            var expected = Input.ReadLines(DAY, $"{name}-answer1")
                ?.FirstOrDefault()
                ?.ToInt32();
            return expected;
        }

        private string GetPart2Expected(string name)
        {
            var expected = Input.ReadLines(DAY, $"{name}-answer2");
            return string.Join("\r\n", expected);
        }


        [Fact]
        public void Part1()
        {
            var input = GetInput("input");
            var expected = GetPart1Expected("input");
            var result = RunPart1(input);

            Assert.True(expected.HasValue);
            Assert.Equal(expected.Value, result);
        }

        [Fact]
        public void Part2()
        {
            var input = GetInput("input");
            var expected = GetPart2Expected("input");
            var compositeImage = BuildCompositImage(input);

            Assert.Equal(expected, compositeImage);
        }

        const int DimensionX = 25;
        const int DimensionY = 6;
        const int ImageSize = DimensionX * DimensionY;


        int RunPart1(List<int> input)
        {
            var foundLayer = GetFindLayerNumberWithLeast(input, 0);

            var layerData = GetImageLayer(input, foundLayer);
            var numOf1s = CountInLayer(layerData, 1);
            var numOf2s = CountInLayer(layerData, 2);

            var result = numOf1s * numOf2s;

            return result;
        }

        int GetFindLayerNumberWithLeast(IList<int> imageData, int value)
        {
            var layerNumber = -1;
            var lowestCount = int.MaxValue;
            var numLayers = imageData.Count / ImageSize;
            for (var i = 0; i < numLayers; i++)
            {
                var layerData = GetImageLayer(imageData, i);
                var count = CountInLayer(layerData, value);
                if (count < lowestCount)
                {
                    layerNumber = i;
                    lowestCount = count;
                }
            }

            return layerNumber;
        }

        IList<int> GetImageLayer(IList<int> imageData, int layerNumber)
        {
            var layerData = imageData
                .Skip(layerNumber * ImageSize)
                .Take(ImageSize)
                .ToList();
            return layerData;
        }
        int CountInLayer(IList<int> layerData, int value)
        {
            var result = layerData.Where(v => v == value).Count();
            return result;
        }

        string BuildCompositImage(IList<int> imageData)
        {
            var sb = new StringBuilder();
            for (var pixel = 0; pixel < ImageSize; pixel++)
            {
                var pixelChar = GetCompositedPixel(imageData, pixel);
                sb.Append(pixelChar);
                if (((pixel + 1) % DimensionX == 0) && (pixel + 1 < ImageSize))
                    sb.AppendLine();
            }
            return sb.ToString();
        }


        char GetCompositedPixel(IList<int> imageData, int pixel)
        {
            var numLayers = imageData.Count / ImageSize;
            for (var layer = 0; layer < numLayers; layer++)
            {
                var idx = (layer * ImageSize) + pixel;
                var value = imageData[idx];
                switch (value)
                {
                    case 0: return ' ';
                    case 1: return '#';
                    default: break;
                }
            }

            return ' ';
        }

    }
}

File: 2019/Python/src/day08.py
import sys
from typing import List


DIMENSION_X:int = 25
DIMENSION_Y:int = 6
IMAGE_SIZE:int = DIMENSION_X * DIMENSION_Y


def part1(input:List[int]) -> int:
    numLayers = get_number_of_layers(input)
    foundLayer = get_layer_number_with_least(input, numLayers, 0)
    layerData = get_image_layer(input, foundLayer)
    num1s = count_in_layer(layerData, 1)
    num2s = count_in_layer(layerData, 2)
    result = num1s * num2s
    return result


def part2(input:List[int]) -> str:
    numLayers = get_number_of_layers(input)
    result = build_composite_image(input, numLayers)
    return result


def get_number_of_layers(imageData:List[int]):
    numLayers = len(imageData) // IMAGE_SIZE
    return numLayers

def get_layer_number_with_least(imageData:List[int], numLayers:int, value:int) -> int:
    layerNumber = -1
    lowestCount = sys.maxsize
    
    for i in range(numLayers):
        layerData = get_image_layer(imageData, i)
        count = count_in_layer(layerData, value)
        if count < lowestCount:
            layerNumber = i
            lowestCount = count

    return layerNumber

def get_image_layer(imageData:List[int], layerNumber:int) -> List[int]:
    start = layerNumber * IMAGE_SIZE
    end = start + IMAGE_SIZE
    layerData = imageData[start:end]
    return layerData

def count_in_layer(imageData:List[int], value:int) -> int:
    matching = list(filter(
        lambda x: (x == value),
        imageData
    ))
    result = len(matching)
    return result

def build_composite_image(imageData:List[int], numLayers:int) -> str:
    result = ''
    for pixelIndex in range(IMAGE_SIZE):
        pixelchar = get_composite_pixel(imageData, numLayers, pixelIndex)
        result += pixelchar
        if (pixelIndex + 1) % DIMENSION_X == 0 and pixelIndex + 1 < IMAGE_SIZE:
            result += '\n'
    return result

def get_composite_pixel(imageData:List[int], numLayers:int, pixelIndex:int) -> str:
    for layer in range(numLayers):
        index = (layer * IMAGE_SIZE) + pixelIndex
        value = imageData[index]
        match value:
            case 0: return ' '
            case 1: return '#'
    return ''